pkg/processor/build/builder.go

```diff
@@ -1552,6 +1520,72 @@ func (b *Builder) getS3FunctionItemKey() (string, error) {
 	return s3Attributes["s3ItemKey"], nil
 }
 
+func (b *Builder) parseFunctionGitCredentials() (*githttp.BasicAuth, error) {
+	cetGitCredentials := b.options.FunctionConfig.Spec.Build.CodeEntryAttributes["gitCredentials"]
+
+	if cetGitCredentials != nil {
+		type GitCredentials struct {
+			Username    string `json:"username,omitempty"`
+			Password    string `json:"password,omitempty"`
+			AccessToken string `json:"accessToken,omitempty"`
+		}
+		var gitCredentials GitCredentials
+
+		if err := mapstructure.Decode(cetGitCredentials, &gitCredentials); err != nil {
+			return nil, errors.Wrap(err, "Failed to decode git credentials map")
+		}
+
+		// use access token if given
+		if gitCredentials.AccessToken != "" {
+			username := "notempty" // if not given use anything but empty (for github - must provide non empty username)
+			if gitCredentials.Username != "" {
+				username = gitCredentials.Username
+			}
+
+			return &githttp.BasicAuth{
+				Username: username,
+				Password: gitCredentials.AccessToken,
+			}, nil
+
+			// otherwise, use username and password authentication if given
+		} else if gitCredentials.Username != "" && gitCredentials.Password != "" {
+
+			return &githttp.BasicAuth{
+				Username: gitCredentials.Username,
+				Password: gitCredentials.Password,
+			}, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func (b *Builder) downloadFunctionFromGit(tempDir, functionPath string) error {
+	var branchRef string
+	var gitAuth *githttp.BasicAuth
+	var err error
+
+	// get branch ref and authorization when given
+	if b.options.FunctionConfig.Spec.Build.CodeEntryAttributes != nil {
+		branchRef, _ = b.options.FunctionConfig.Spec.Build.CodeEntryAttributes["branch"].(string)
+		if gitAuth, err = b.parseFunctionGitCredentials(); err != nil {
+			return errors.Wrap(err, "Failed to parse function git credentials")
+		}
+	}
+
+	if _, err = git.PlainClone(tempDir, false, &git.CloneOptions{
+		URL:           functionPath,
+		ReferenceName: plumbing.ReferenceName(branchRef),
```

Comment:
that won't work for
1. tags
2. azure devops

1. to get tags support you

```go
repo, _ := git.PlainClone(...)
revision, err := repository.ResolveRevision(plumbing.Revision(referenceName))
worktree, _ := git.Worktree(...)
worktree.Checkout(&git.CheckoutOptions{Hash:...})
```

2. must give a workaround, perhaps using shell command, something like:

`git -c http.extraHeader="Authorization: Basic ${B64_PAT}" clone https://dev.azure.com/<username>/nuclio/_git/nuclio --branch main --single-branch`